server/src/storage/mod.rs

```diff
@@ -13,101 +13,117 @@
  * limitations under the License.
  */
 
-//! Converts raw MQTT messages to uncompressed data points, stores uncompressed data points
-//! temporarily in an in-memory buffer that spills to Apache Parquet files, and stores data points
-//! compressed as models in Apache Parquet files.
+//! Converts raw MQTT messages to uncompressed data points, stores uncompressed data points temporarily
+//! in an in-memory buffer that spills to Apache Parquet files, and stores data points compressed as
+//! models in memory to batch compressed data before saving it to Apache Parquet files.
 
 mod data_point;
 mod segment;
+mod time_series;
 
 use std::collections::vec_deque::VecDeque;
 use std::collections::HashMap;
-use std::fs;
 use std::fs::File;
 
 use datafusion::arrow::record_batch::RecordBatch;
 use datafusion::parquet::arrow::ArrowWriter;
 use datafusion::parquet::basic::Encoding;
 use datafusion::parquet::file::properties::WriterProperties;
 use paho_mqtt::Message;
-use tracing::{error, info, info_span};
+use tracing::{info, info_span};
 
 use crate::storage::data_point::DataPoint;
 use crate::storage::segment::{FinishedSegment, SegmentBuilder};
-use crate::types::Timestamp;
+use crate::storage::time_series::CompressedTimeSeries;
+
+// TODO: Look into moving handling of uncompressed and compressed data into separate structs.
+// TODO: Look into custom errors for all errors in storage engine.
 
 // Note that the initial capacity has to be a multiple of 64 bytes to avoid the actual capacity
 // being larger due to internal alignment when allocating memory for the builders.
 const INITIAL_BUILDER_CAPACITY: usize = 64;
-const RESERVED_MEMORY_IN_BYTES: usize = 5000;
+const UNCOMPRESSED_RESERVED_MEMORY_IN_BYTES: usize = 5000;
+const COMPRESSED_RESERVED_MEMORY_IN_BYTES: isize = 5000;
```

Comment:
The main reason is that for uncompressed data we can check the remaining memory before allocating space since we first actually use the memory when creating a builder. For compressed data the memory is occupied as soon as the compressed segment is received so instead of checking the remaining memory before inserting, we instead insert it first and check if the remaining memory is below 0 after. 

This should be okay to do briefly since the memory is allocated for the compressed segment already anyway and can therefore be seen as being part of the memory overhead of running the storage engine. 

I have added a comment to the const that tries to explain this.

Reply:
The alternative would be to check the remaining bytes before inserting the compressed data which is also an option. This does make the structure of `insert_compressed_data` a bit more confusing. However, I think the structure needs to change when adding the `MAX_COMPRESSED_FILE_SIZE` setting later anyway, so i can look into this issue again then.